docs/apache-airflow/administration-and-deployment/public-airflow-interface.rst

```diff
@@ -0,0 +1,87 @@
+
+You can extend Airflow in three ways:
+
+* By writing new custom Python code (via Operators, Plugins, Provider)
+* By using the `Stable REST API <stable-rest-api-ref>`_ (based on the OpenAPI specification)
+* By using the `Airflow Command Line Interface (CLI) <cli-and-env-variables-ref.rst>`_
+
+How can you extend Apache Airflow with custom Python Code?
+==========================================================
+
+The Public Interface of Airflow consists of a number of different classes and packages that provide access
+to the core features and functionality of the system.
+
+The classes and packages that may be considered as the Public Interface include:
+
+* The :class:`~airflow.DAG`, which provides a way to define and manage DAGs in Airflow.
+* The :class:`~airflow.models.baseoperator.BaseOperator`, which provides a way write custom operators.
+* The :class:`~airflow.hooks.base.BaseHook`, which provides a way write custom hooks.
+* The :class:`~airflow.models.connection.Connection`, which provides access to external service credentials and configuration.
+* The :class:`~airflow.models.variable.Variable`, which provides access to Airflow configuration variables.
+* The :class:`~airflow.models.xcom.XCom` which are used to access to inter-task communication data.
+* The :class:`~airflow.secrets.BaseSecretsBackend` which are used to define custom secret managers.
+* The :class:`~airflow.plugins_manager.AirflowPlugin` which are used to define custom plugins.
+* The :class:`~airflow.triggers.base.BaseTrigger`, which are used to implement custom Custom Deferrable Operators (based on ``asyncio``).
+* The :class:`~airflow.decorators.base.TaskDecorator`, which provides a way write custom decorators.
+* The :class:`~airflow.listeners.listener.ListenerManager` class which provides hooks that can be implemented to respond to DAG/Task lifecycle events.
+
+.. versionadded:: 2.5
+
+   Listener public interface has been added in version 2.5.
+
+* The :class:`~airflow.executors.base_executor.BaseExecutor` - the Executors are the components of Airflow
+  that are responsible for executing tasks.
+
+.. versionadded:: 2.6
+
+   There are a number of different executor implementations built-in Airflow, each with its own unique
+   characteristics and capabilities. Executor interface was available in earlier version of Airflow but
+   only as of version 2.6 executors are fully decoupled and Airflow does not rely on built-in set of executors.
+   You could have implemented (and succeeded) with implementing Executors before Airflow 2.6 and a number
+   of people succeeded in doing so, but there were some hard-coded behaviours that preferred in-built
+   executors, and custom executors could not provide full functionality that built-in executors had.
+
+
+What is not part of the Public Interface of Apache Airflow?
+===========================================================
+
+Everything not mentioned in this document should be considered as non-Public Interface.
```

Comment:
This doesn't address providers, or how we can include "private" e.g. helper code in otherwise "public" modules / classes.

Is it sufficient to include `:meta private:`?  Or must we prefix with underscore?

Reply:
IMHO - it should also address providers (and especially them). 

Rather than adding private tags, I think we should allowlist everything that should be public here. If we forgot something we should add.  All the rest should be private by definition IMHO.  This is the whole purpose of this list. And we should explicitly add here everything, all helper methods that providers are supposed to use (and add some automation to be able to enable verifying it - for example using the same approach we do in common.sql now using stubgen).

Otherwise we are not really defining public interface.

Reply:
And the list does not **have** to be complete now. If we forget something before merging it - we can always add it later - there is completely no harm in that. It will be "eventually complete" if we follow it for quite some time and decide that some things should be "publicised".

Reply:
And yes I want to add eventually tests to make sure that all our providers are only using what we consider public" and nothing else. This will be partially done as part of AIP-44 (for DB methods) but it can be extended to all Python calls made by the providers.

Reply:
Sorry... so ... specifically in the provider package... is everything private, public, or some private some public?

Reply:
And if it's both then how do we distinguish?  I think the current convention is underscores OR meta private tag.

But interested in your thoughts.  We should include that in this proposal i think.

Personally I think it is an important goal to minimize backcompat surface area so i would suggest that everything that's not an operator / sensor / hook should be private (i.e. all utils / helper code)

Reply:
It's also important to clarify that some methods are intentionally "private". E.g. in the operator, it might have helper methods.  But these methods are more implementation detail right?  If users subclass, though, they may use and depend on these.  So what do we do? Ideally, from maintainer perspective, we want the overall behavior of the operator to be subject to backcompat but not all of its methods.  Currently the convention is, i believe, all methods are assumed public unless prefix with underscore / marked meta private.  WDYT?

Reply:
One problem with public / private is... from tool perspective ... even using _my_func across modules is forbidden.   But there is a difference between the internal "private" and the user-facing "private".  We need to be able to write helper code that can be used across modules (internal public), but not have it be user-facing public.

Reply:
Maybe this is misunderstanding - what I wrote above was not about what provider packages "provides" - but what they "consume". This  document indeed does not specify anything about what providers provide - it's only what Airflow "produces" and providers "consume" as users of that. 

The Community Providers have their own already pretty very well defined interface (with defined json schema) for public API they provide (introduced since the beginning in Airflow 2.0): provider.yaml kept internally and corresponding public `get_provider_info` entrypoint which explains all packages/classes/functionalities that are exposed and also `airflow providers` CLI shows that information. 

I think (maybe I am wrong) that tverything specified there is public - including automatically generated list of "core extensions" the provider deliver - and the documentation about that is automatically updated based on this meta-data (and things like naming, completeness etc. is automatically verified at PR time with unit tests following AIP-21).

And yes. In Providers, whenever something is ment to be internal there is indeed defined with `_` and our sphinx docs building excludes it automatically.  Actually all those are things we cross-check against when we decide on making a breaking release of the providers or adding new features. And I think it's very well controlled by the pre-commits/cross-verification and general very limited "code structure" we have there - hooks, contain hooks, operators contain operators etc. 

But I think this  is very much different for Airlow "core" - there are multiple classes that have no `_` and they are not really specified if they could or should be used by the users to rely on and not marked with :private:  nor _. And I think keeping that in check could be difficult. 

But maybe that is a good idea? Do you think @dstandish that we should synchronize and make similar approach for airflow - instead of allowlisting and explicitly specifying what are the public classes we should use,  to make an effort to review all those Airlfow classes in airflow and rename them to _* for example to follow what has been done and followed in providers?

 I would also be very happy if we would do it (and we could even add pre-commit to add some friction to add a new class that is public classes).  But I think this might be pretty invasive for multiple reasons.

Reply:
> It's also important to clarify that some methods are intentionally "private". E.g. in the operator, it might have helper methods. But these methods are more implementation detail right? If users subclass, though, they may use and depend on these. So what do we do? Ideally, from maintainer perspective, we want the overall behavior of the operator to be subject to backcompat but not any of its methods. Currently the convention is, i believe, all methods are assumed public unless prefix with underscore / marked meta private. WDYT?

For providers it's far easier because there is no "internal" sharing between multiple modules. Each of them is really either "standalone" or uses "public" classes from other providers. And with common.sql we even introduced stubgen to generate the actual "public interface" that MyPy can surface as errors. And we even have a friction added (necessiety of regenerating the stubs) when someone changes the API by accident.

> One problem with the public / private distinction is... from tool perspective ... even using _my_func across modules is sorta forbidden. But there is a difference between the internal "private" and the user-facing "private". We need to be able to write helper code that can be used across modules (internal public), but not have it be user-facing public.

This is very right assesment.

That's why I proposed to allowlist and be explicit about everything that is public. And from the tool perspective, one thing is  important (what we did for common.sql) - to keep the list of all public methods, and whenever something there is *removed* or *updated*, fail the build and make it an explicit action (friction) for the one who modifies it accidentally. We can also add a tool for whoever uses airflow tools to check if they are not using something not "intentionally" public (we've done that for common.sql - via MyPy stubs).

But we cannot  (yet) do any automation in Airflow, because first we need to agree and document what is and what is not public. We need a good list of things that should be "public" - so that we can (eventually) add friction to removing or updating those and possibly eventually add a tool for others to verify if they are not using something they should not. 

This doc here is mostly to start the discussion and work-out a good set of classes/packages that we should consider "really public". For now in a document, eventually "guarded automatically".

BTW. As I mentioned in many places, we are not able to get it 100% correct. Never. [Hyrum's law](https://www.hyrumslaw.com/) is very clear and IMHO very true about this. 

But we can state our intentions, and make an effort to keep those "explicitly intended" APIs controlled. 

And first ... we need to agree on what our intentions are. If we don't document them and don't agree what is and what is not our intention for public API - everyone will have their own understanding of it.  And often those understandings will be different one (for example many users still use the DB of Airflow and relies on its structure because they assume they can rely on it). You wrote yourself that you prefer to `minimize backcompat surface` and "everything that's not an operator / sensor / hook should be private". But this is only your understanding - others might say "everything in utils can be used by DAG authors and provider writers".  Both statement are possible - both orthogonally different. Both might be deliberate decision we should make as a community. But we need to make that decision and be very explicit about it. This is a starting point that is missing now IMHO.

I just want to have a page where we all agree what we intentionally make "public", and have clear rules that all the rest is not.

Reply:
BTW. Eventually we might be able to produce airflow typeshed https://github.com/python/typeshed with only stubs for only "externally public" APIs. This is what should be an "ultimate" goal I think.

Reply:
> you wrote yourself that you prefer to minimize backcompat surface and "everything that's not an operator / sensor / hook should be private".

just want to clarify that, specifically with that example about "only hook / sensor / operator are public" i was not quite saying that's what i think it should be, but speaking to the ideal from the perspective of a maintainer.  strictly from maintainer hat, your job is easier the less backcompat you have to worry about.

Reply:
> For providers it's far easier because there is no "internal" sharing between multiple modules

so i mean even e.g. for KPO we have a `pod_manager` module.  This should not be considered (ideally) user-public.  It exists only for KPO, used only in KPO.  It is internal helper code for KPO, just because, otherwise, KPO is 1000 line class.  (maybe exaggerating, maybe not). But it should be internal-public.  So there is sharing, and ideally, from maintainer perspective this module isn't user-public.  And moreover, again from this perspective, ideally operator methods should not be user-public.  Makes it harder to refactor and make changes.  The result here would be, "subclass operators at your own risk".  Would this be acceptable?  I feel like it could be OK.  So then "base" operator / sensor etc could be the public API (at least, whatever's not explicitly private there via meta tag or _), and then any _particular_ operator's methods are not public?

Reply:
That's a good point. I think K8S is pretty special case because it is both used internally by K8S executor as well as it is exposed externally. 

And I think there we **could** treat it in a special way and just there make sure that we explicitly decide on which parts of the pod manager or KPO is bound to change. We could indeed decide that the _ (or :private:) is actually  something that you SHOULD NOT use because it MIGHT change.  And I think in most cases we follow that convention actually  - I see tht in KPO we have a number of classes that are `_` and yes - that's ok for those methods to change their signatures in whatever way. User got enough signal (and we should be explicit about it too) that this method is more volatile than the others (by sheer _ in front).

And to be honest - this is already supported by our documentation: https://airflow.apache.org/docs/apache-airflow-providers-cncf-kubernetes/stable/_api/airflow/providers/cncf/kubernetes/index.html  - does not contain any _* methods because they are excluded by our Sphinx autoapi settings.

I will add some description about that to cover general approach for that (also covering providers). 

Also I have found one thing. We actually have **almost** what I created here available (but no-one commenting here realized that so I consider that as something that we can easily merge.

We have this page in our docs: https://airflow.apache.org/docs/apache-airflow/stable/python-api-ref.html -  and besides being out-dated and having some buggy entries (empty "utils" ?) it was an attempt to do very similar as part of my doc - expose some of the classes that we think are "public". And besides being unmaintained (no triggers for example) It had no explicit mentioning that this is "public interface" and all the rest is not and it was only about "Python API", but Airflow public interface is a little more than that (DB/REST Api and explicit exclusion of the UI and possibly other components are more general than just Python API). I will remove that page and add redirection and merge these two.

Reply:
Yeah, just not only kpo, I think there are other providers with e.g. utils.py